service/frontend/workflow_handler.go

```diff
@@ -3108,6 +3111,246 @@ func (wh *WorkflowHandler) validateStartWorkflowArgsForSchedule(
 	return wh.validateSearchAttributes(unaliasedStartWorkflowSas, namespaceName)
 }
 
+func (wh *WorkflowHandler) DescribeDeployment(ctx context.Context, request *workflowservice.DescribeDeploymentRequest) (_ *workflowservice.DescribeDeploymentResponse, retError error) {
+	defer log.CapturePanic(wh.logger, &retError)
+
+	if request == nil {
+		return nil, errRequestNotSet
+	}
+
+	if len(request.Namespace) == 0 {
+		return nil, errNamespaceNotSet
+	}
+
+	if !wh.config.EnableDeployments(request.Namespace) {
+		return nil, errDeploymentsNotAllowed
+	}
+
+	namespaceID, err := wh.namespaceRegistry.GetNamespaceID(namespace.Name(request.GetNamespace()))
+	if err != nil {
+		return nil, err
+	}
+
+	// validating params
+	maxIDLengthLimit := wh.config.MaxIDLengthLimit()
+	err = deployment.ValidateDeploymentWfParams(deployment.SeriesFieldName, request.Deployment.SeriesName, maxIDLengthLimit)
+	if err != nil {
+		return nil, err
+	}
+	err = deployment.ValidateDeploymentWfParams(deployment.BuildIDFieldName, request.Deployment.BuildId, maxIDLengthLimit)
+	if err != nil {
+		return nil, err
+	}
+
+	deploymentWorkflowID := deployment.GenerateDeploymentWorkflowID(request.Deployment.SeriesName, request.Deployment.BuildId)
+
+	req := &historyservice.QueryWorkflowRequest{
+		NamespaceId: namespaceID.String(),
+		Request: &workflowservice.QueryWorkflowRequest{
+			Namespace: request.Namespace,
+			Execution: &commonpb.WorkflowExecution{
+				WorkflowId: deploymentWorkflowID,
+			},
+			Query: &querypb.WorkflowQuery{QueryType: deployment.QueryDescribeDeployment},
+		},
+	}
+
+	res, err := wh.historyClient.QueryWorkflow(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var queryResponse deployspb.DescribeResponse
+	err = payloads.Decode(res.GetResponse().GetQueryResult(), &queryResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	// build out task-queues for the response object
+	var taskQueues []*deploypb.DeploymentInfo_TaskQueueInfo
+	deploymentLocalState := queryResponse.DeploymentLocalState
+
+	for taskQueueName, taskQueueFamilyInfo := range deploymentLocalState.TaskQueueFamilies {
+		for _, taskQueueInfo := range taskQueueFamilyInfo.TaskQueues {
+			element := &deploypb.DeploymentInfo_TaskQueueInfo{
+				Name:            taskQueueName,
+				Type:            enumspb.TaskQueueType(taskQueueInfo.TaskQueueType),
+				FirstPollerTime: taskQueueInfo.FirstPollerTime,
+			}
+			taskQueues = append(taskQueues, element)
+		}
+	}
+
+	return &workflowservice.DescribeDeploymentResponse{
+		DeploymentInfo: &deploypb.DeploymentInfo{
+			Deployment:     deploymentLocalState.WorkerDeployment,
+			CreateTime:     deploymentLocalState.CreateTime,
+			TaskQueueInfos: taskQueues,
+			Metadata:       deploymentLocalState.Metadata,
+			IsCurrent:      deploymentLocalState.IsCurrentDeployment,
+		},
+	}, nil
```

Comment:
I agree on consolidating the deployment workflow client. However, I still would like to play with the idea that the client should be namespace + deployment specific rather than pass these names in separate function calls. Even in the frontend, from a quick gauge, it seems like we have all the required information for instantiating the client. I will be diving deeper into this right now, but pasting this to see if you had any more reasons for not going ahead with this approach